### ISSUE TITLE
Add PyTorch version of actor-critic Coursera assignment

### ISSUE THREAD

**Issue by dniku** [yandexdataschool/Practical_RL#398]
We have `week5_policy_based/practice_a3c.ipynb` only in Tensorflow since it does not have a corresponding assignment in `master`.

There are multiple issues with this notebook. For one, it does not implement A3C, only a basic actor-critic. It is also a heavily modified version of master/week08/practice_pomdp which was never originally intended to be an actor-critic assignment.

**Comment by AI-Ahmed** [comment]
Have you done anything so far about this problem? Besides, I am facing issues in the assignment of week08 related to this issue. I am trying to fix both now; the policy loss and the reward are not correct in both studies (week08 and week06), although I have done everything that I can be done to fix both of them. I want to know if it is something regards the` atari_util.py` file and not in our code!!!

**Comment by AI-Ahmed** [comment]
Some screenshots of both cases;

Week 06:
![image](https://user-images.githubusercontent.com/72295771/201522360-ca9bfbe7-2c6c-4682-ab6f-ab29624f0f3f.png)

Week 08:

![image](https://user-images.githubusercontent.com/72295771/201522393-684d9693-4513-4583-ace8-ad326d518a79.png)

**Comment by dniku** [comment]
@AI-Ahmed 

We haven't done anything about this assignment; but this issue is about the Coursera assignment specifically, and not the ones in the `master` branch, which, I assume, you are talking about.

Your screenshots of plots seem to indicate that your agent isn't learning anything at all, and is behaving randomly. I'd guess that the reason is some bug in your code, e.g. a `-` sign missing before the loss. You may want to refer to some open-source implementation of A2C — e.g. [this one](https://github.com/DLR-RM/stable-baselines3/blob/7e1db1aaaa0f486cceb79faf5a08a25c5fded670/stable_baselines3/a2c/a2c.py#L149-L162) — to compare it with yours and possibly spot some errors.

**Comment by AI-Ahmed** [comment]
It is weird @dniku , because I think I already did that correctly!!

Week 06:
![image](https://user-images.githubusercontent.com/72295771/201537002-be17d7fd-b090-4c31-a497-97f1d8ff1c2b.png)

Week 08:

- Without adding λ
```python
  logits = torch.stack(logits, dim=1)
    state_values = torch.stack(state_values, dim=1)
    probas = F.softmax(logits, dim=2)
    logprobas = F.log_softmax(logits, dim=2)

    # select log-probabilities for chosen actions, log pi(a_i|s_i)
    actions_one_hot = to_one_hot(actions, n_actions).view(
        actions.shape[0], actions.shape[1], n_actions)
    logprobas_for_actions = torch.sum(logprobas * actions_one_hot, dim=-1)

    # Now let's compute two loss components:
    # 1) Policy gradient objective.
    # Notes: Please don't forget to call .detach() on advantage term. Also please use mean, not sum.
    # it's okay to use loops if you want
    J_hat = 0  # policy objective as in the formula for J_hat

    # 2) Temporal difference MSE for state values
    # Notes: Please don't forget to call on V(s') term. Also please use mean, not sum.
    # it's okay to use loops if you want
    value_loss = 0

    cumulative_returns = state_values[:, -1].detach()

    for t in reversed(range(rollout_length)):
        r_t = rewards[:, t]                                # current rewards
        # current state values
        V_t = state_values[:, t]
        V_next = state_values[:, t + 1].detach()           # next state values
        # log-probability of a_t in s_t
        logpi_a_s_t = logprobas_for_actions[:, t]

        # update G_t = r_t + gamma * G_{t+1} as we did in week6 reinforce
        cumulative_returns = G_t = r_t + gamma * cumulative_returns

        # Compute temporal difference error (MSE for V(s))
        approx_A_t = r_t.unsqueeze(1) + gamma * V_next
        value_loss += F.mse_loss(approx_A_t, V_t).detach() #<YOUR CODE>

        # compute advantage A(s_t, a_t) using cumulative returns and V(s_t) as baseline
        advantage = cumulative_returns -  V_t #<YOUR CODE>
        advantage = advantage.detach()

        # compute policy pseudo-loss aka -J_hat.
        J_hat += (logpi_a_s_t * advantage).detach().mean() #<YOUR CODE>

    # regularize with entropy
    entropy_reg = (logprobas * probas).sum(dim=-1).mean() #<YOUR CODE: compute entropy regularizer>

    # add-up three loss components and average over time
    loss = -J_hat + value_loss + -0.01 * entropy_reg
```

- With adding λ [0.92] --> **[(John Schulman et al, 2015)](https://arxiv.org/abs/1506.02438)**
```python
  logits = torch.stack(logits, dim=1)
    state_values = torch.stack(state_values, dim=1)
    probas = F.softmax(logits, dim=2)
    logprobas = F.log_softmax(logits, dim=2)

    # select log-probabilities for chosen actions, log pi(a_i|s_i)
    actions_one_hot = to_one_hot(actions, n_actions).view(
        actions.shape[0], actions.shape[1], n_actions)
    logprobas_for_actions = torch.sum(logprobas * actions_one_hot, dim=-1)

    # Now let's compute two loss components:
    # 1) Policy gradient objective.
    # Notes: Please don't forget to call .detach() on advantage term. Also please use mean, not sum.
    # it's okay to use loops if you want
    J_hat = 0  # policy objective as in the formula for J_hat

    # 2) Temporal difference MSE for state values
    # Notes: Please don't forget to call on V(s') term. Also please use mean, not sum.
    # it's okay to use loops if you want
    value_loss = 0

    cumulative_returns = state_values[:, -1].detach()

    for t in reversed(range(rollout_length)):
        r_t = rewards[:, t]                                # current rewards
        # current state values
        V_t = state_values[:, t]
        V_next = state_values[:, t + 1].detach()           # next state values
        # log-probability of a_t in s_t
        logpi_a_s_t = logprobas_for_actions[:, t]

        # update G_t = r_t + gamma * G_{t+1} as we did in week6 reinforce
        cumulative_returns = G_t = r_t + gamma * cumulative_returns

        # Compute temporal difference error (MSE for V(s))
        approx_A_t = r_t.unsqueeze(1) + gamma * V_next
        value_loss += F.mse_loss(approx_A_t, V_t).detach() #<YOUR CODE>

        # compute advantage A(s_t, a_t) using cumulative returns and V(s_t) as baseline
        advantage = lamda * cumulative_returns - (1-lamda) * V_t #<YOUR CODE>
        advantage = advantage.detach()

        # compute policy pseudo-loss aka -J_hat.
        J_hat += (logpi_a_s_t * advantage).detach().mean() #<YOUR CODE>

    # regularize with entropy
    entropy_reg = (logprobas * probas).sum(dim=-1).mean() #<YOUR CODE: compute entropy regularizer>

    # add-up three loss components and average over time
    loss = -J_hat + value_loss + -0.01 * entropy_reg

    # Gradient descent step
    # <YOUR CODE>
    opt.zero_grad()
    loss.backward()
    # Clip the gradients for gradient's normalization
    nn.utils.clip_grad_norm_(agent.parameters(), 0.5)
    opt.step()
```

**Comment by AI-Ahmed** [comment]
The problem was solved in both notebooks. The problem I had in both notebooks was related to multiplying `is_not_done` by the `value_target`.  When I didn't do that, that added more potential to the imperfect value function rather than adding the possibility to the reward, and the agent thought that there was no end to the episode, which means that all the value functions for all the states have equality due to the continuouaty.
